cw/cw305/cw_lite_segmented.py

```diff
@@ -0,0 +1,175 @@
+
+
+class CwLiteSegmented:
+    """Class for capturing traces using a ChipWhisperer-Lite.
+
+    This class uses segmented traces mode to improve capture performance.
+
+    When in segmented mode, ChipWhisperer-Lite captures multiple waves each starting at
+    a trigger event. This is much more efficient than sending a separate command for
+    each wave.
+
+    Due to current limitations of the hardware, number of waves (segments) per capture
+    is fixed and there is also a lower limit on the number of samples per wave. This
+    class handles this internally but also makes this information available via the
+    read-only ``num_segments_actual`` attribute so that the target can be configured to
+    generate the correct number of triggers.
+
+    Typical usage:
+    >>> cw_lite = CwLiteSegmented()
+    >>> while foo:
+    >>>     ...
+    >>>     cw_lite.num_segments = desired_num_segments
+    >>>     # Note: cw_lite.num_segments_actual gives the actual number of segments that
+    >>>     # will be captured.
+    >>>     cw_lite.arm()
+    >>>     # Configure the target using cw_lite.num_segments_actual.
+    >>>     target.configure(cw_lite.num_segments_actual)
+    >>>     # This returns cw_lite.num_segments number of waves each with
+    >>>     # cw_lite.num_samples samples.
+    >>>     waves = cw_lite.capture_and_transfer_waves()
+    >>>     ...
+
+    Attributes:
+        num_samples_min: Minimum number of samples per wave. Read-only.
+        num_samples_max: Maximum number of samples per wave. Read-only.
+        num_samples: Number of samples per wave, must be in [``num_samples_min``,
+            ``num_samples_max``].
+        num_segments_min: Minimum number of waves per capture. Read-only.
+        num_segments_max: Maximum number of waves per capture, depends on the number of
+            samples per wave. Read-only.
+        num_segments: Number of waves per capture, must be in [``num_segments_min``,
+            ``num_segments_max``].  This number determines the number of waves returned
+            by ``capture_and_transfer_waves``.
+        num_segments_actual: Actual number of waves that will be captured. This number
+            depends on the number of samples per wave. The target must create this many
+            trigger events. Read-only.
+    """
+
+    def __init__(self, num_samples=740):
+        """Inits a CwLiteSegmented.
+
+        Args:
+            ip_addr: IP address of the oscilloscope.
+        """
+        self._scope = cw.scope()
+        self._configure_scope()
+        self.num_segments = 1
+        self.num_samples = num_samples
+        self._print_device_info()
+
+    @property
+    def num_segments_min(self):
+        return 1
+
+    @property
+    def num_segments_max(self):
+        return (self._scope.adc.oa.hwMaxSamples // self._scope.adc.samples) - 1
+
+    @property
+    def num_segments_actual(self):
+        # Must round-up to fill the entire buffer.
```

Comment:
This is not clear to me but I think I now start to understand.

It seems to me that we actually want to capture `num_segments_max` segments per batch which with the current settings would be floor(24573/742) - 1 = 32. This is the maximum number of complete segments that CW-Lite can hold in its sample buffer.

If I understand it correctly, we must - due to some limitations in the firmware - always fill the complete sample buffer. I.e., after the 32 segments we collect more samples until the buffer is full. Since this last segment is not captured completely, we will discard it.

Is that correct? if yes, it would be great if you could clarify a bit the comments.

Reply:
Yes, your understanding is correct but let me rephrase for the sake of clarity: CW-Lite always captures `num_segments_actual` segments, this depends on the number of samples per trace and the target must create this many trigger events. `num_segments` can be anything in `[num_segments_min, num_segments_max]` and is basically a convenience so that we can encapsulate post-processing in this class.

Can you take another look at the docstring of the class (I thought that along with this comment was clear enough)?

Reply:
Thanks for confirming. I did read the docstring but it wasn't clear to me. I have now made a suggestion how to clarify the docstring.